lib/services/database/comment_repository_service.dart

```diff
@@ -0,0 +1,124 @@
+/// of the posts in the firestore database.
+class CommentRepositoryService {
+  final FirebaseFirestore _firestore;
+
+  CommentRepositoryService({required FirebaseFirestore firestore})
+      : _firestore = firestore;
+
+  /// Returns the document reference of the post with id [postId]
+  DocumentReference<Map<String, dynamic>> _postDocument(
+    PostIdFirestore postId,
+  ) {
+    return _firestore
+        .collection(PostFirestore.collectionName)
+        .doc(postId.value);
+  }
+
+  /// Returns the collection reference of the subcollection of comments of
+  /// the post with id [parentPostId]
+  CollectionReference<Map<String, dynamic>> _commentsSubCollection(
+    PostIdFirestore parentPostId,
+  ) {
+    return _postDocument(parentPostId)
+        .collection(CommentFirestore.subCollectionName);
+  }
+
+  /// This method returns the comments of the post with id [parentPostId]
+  Future<List<CommentFirestore>> getComments(
+    PostIdFirestore parentPostId,
+  ) async {
+    final commentsQuery = await _commentsSubCollection(parentPostId).get();
+
+    final comments = commentsQuery.docs
+        .map((docSnap) => CommentFirestore.fromDb(docSnap))
+        .toList();
+
+    return comments;
+  }
+
+  /// This method will add the comment with data [commentData] to the
+  /// post with id [parentPostId].
+  /// It will also update the number of comments of the post.
+  /// This is done atomically.
+  ///
+  /// The method returns the id of the comment that was added.
+  Future<CommentIdFirestore> addComment(
+    PostIdFirestore parentPostId,
+    CommentData commentData,
+  ) async {
+    // Generate a new reference for the comment
+    // Although generated locally, the new id can be considered unique
+    // https://stackoverflow.com/questions/54268257/what-are-the-chances-for-firestore-to-generate-two-identical-random-keys
+    final newCommentRef = _commentsSubCollection(parentPostId).doc();
+
+    // Create a batch write to perform the operations atomically
+    final batch = _firestore.batch();
+
+    batch.set(newCommentRef, commentData.toDbData());
+
+    final postDocRef = _postDocument(parentPostId);
+    batch.update(
+      postDocRef,
+      {PostData.commentCountField: FieldValue.increment(1)},
+    );
+
+    await batch.commit();
+
+    return CommentIdFirestore(value: newCommentRef.id);
+  }
+
+  /// This method will delete the comment with id [commentId] from the
+  /// post with id [parentPostId].
+  /// It will also update the number of comments of the post.
+  /// This is done atomically.
+  /// If the comment does not exist, the method will do nothing.
+  Future<void> deleteComment(
+    PostIdFirestore parentPostId,
+    CommentIdFirestore commentId,
+  ) async {
+    await _firestore.runTransaction((transaction) async {
+      await _deleteComment(parentPostId, commentId, transaction);
+    });
+  }
+
+  // Concrete implementation of the deletion of a comment
+  Future<void> _deleteComment(
```

Comment:
I am not sure why we have a private method here, as it is only used once in `deleteComment`.
I assume it is to avoid a lengthy async transaction body, but in my opinion the locality of behavior principle should be applied here.

Reply:
This was indeed to reduce the body of the run transation and to have a better separation of logic.
I feel like it makes the code easier to read.

Also, I just noticed that using this function I could use the arrow notation to have the `deleteComment` more consice. I do that in 2d01850c7f0221d6121ec633d3c6587df1546db6.

Please tell me what you think about it and if you think it's still better to put it in the transaction body then I'll do it :+1:

Reply:
It might be a personal opinion, but I don't really like this syntax :rofl:...

I prefer to keep the function body and have everything in the transaction for the reason explained above. It is not very important if you feel strongly about this.

Reply:
Ok I see :sweat_smile: .
Then I'll move it in the body of `runTransaction` (this actually makes the code more concise).
Done in 7d5766d3590391cfae2ffc7898fa7a480b4cb89c.